### DISCUSSION TITLE
Make one-to-many relation on user->role?

### DISCUSSION THREAD

**Discussion by xhava** [spatie/laravel-permission#1786] [Q&A]
hi, i would like to make a one of many Polymorphic Relationship 
i just want to get just one role per user, and not to many.

it is posible with the One of May Relation?

o would like to make a currentRole method and get just one role and not a collection of many roles. 

i would like to achieve this functionality in the relation and not in another method to filter the roles and get the first. im trying to make in the relation, but i don`t know if the "One of Many" relation could do this.

Thanks for the help.

```
public function roles(): BelongsToMany
    {
        return $this->morphToMany(
            config('permission.models.role'),
            'model',
            config('permission.table_names.model_has_roles'),
            config('permission.column_names.model_morph_key'),
            'role_id'
        );
    }
```

**Comment by xhava** [comment]
i made this, and i am not shure that is the best solution.
1.- add current_role to the appends
```
protected $appends = [
        'profile_photo_url',
        'current_role',
    ];
```

2.- set the first role of the relation "roles" in the HasRoles Trait.

```
public function getCurrentRoleAttribute()
    {
        return $this->roles()->first();
    }
```


some one that can help me to check if this is a good option or not. thanks a lot.

**Reply by erikn69** [reply]
It's not the best, It has performance cost
I propose these changes
```php
public function getCurrentRoleAttribute()
{
    return ($this->relations['roles'] ?? false) ? ($this->relations['roles'][0] ?? null) : null;
}
```
So, when you need `current_role` must use
```php
\App\Models\User::with(['roles'=>function($q){ $q->groupBy('user_id'); }])->get()
```
`with('roles')` avoid to call for every user, `$q->groupBy('user_id');` get only one role per user, if there are no `with('roles')` avoid getting unnecesary data from db and `current_role=null`

**Reply by xhava** [reply]
really thanks for the suggestion and for take some of your time on answer.
i´ will make the update with your suggestion. 

thanks a lot.
